src/mlpack/methods/kde/kde_rules_impl.hpp

```diff
@@ -76,16 +78,58 @@ double KDERules<MetricType, KernelType, TreeType>::BaseCase(
   if ((lastQueryIndex == queryIndex) && (lastReferenceIndex == referenceIndex))
     return 0.0;
 
-  // Calculations.
-  const double distance = metric.Evaluate(querySet.col(queryIndex),
-                                          referenceSet.col(referenceIndex));
-  densities(queryIndex) += kernel.Evaluate(distance);
+  // Use PCA evaluation if possible.
+  if (pca)
+  {
+    const KDEStat& stat = traversalInfo.LastReferenceNode()->Stat();
+
+    const arma::vec& qPoint = querySet.col(queryIndex);
+    const arma::vec qProj = stat.EigVec().t() * (qPoint - stat.Mean());
+    const arma::vec qRecon = stat.EigVec() * qProj + stat.Mean();
+    const double kernelValue1 = EvaluateKernel(qPoint, qRecon);
+
+    // Base case for each reference node.
+    const arma::vec& rPoint = referenceSet.col(referenceIndex);
+    const arma::vec rProj = stat.EigVec().t() * (rPoint - stat.Mean());
+    const arma::vec rRecon = stat.EigVec() * rProj + stat.Mean();
+
+    const double maxMetricError = metric.Evaluate(rRecon, rPoint);
```

Comment:
There are two sources of error: the reference point reconstruction error, which you have here as `maxMetricError` and then the query point reconstruction error (i.e., `metric.Evaluate(qRecon, qPoint)`).  So when we compute `maxProjKernel`, we should use `kernel.Evaluate(metricProjValue - maxMetricError - maxQueryMetricError)` (also need to make sure that's nonnegative), and for `minProjKernel` we should use `kernel.Evaluate(metricProjValue + maxMetricError + maxQueryMetricError)`.

I suspect that all the extra overhead of computing these projections will actually not help save any runtime; but, if we can get this to work, we can then figure out the best way to incorporate it into the tree pruning rules, and that is where we can make it fast. :)

Reply:
I've tried this but `(metricProjValue - maxMetricError - maxQueryMetricError) < 0` in every case (using `KDETest/GaussianSingleKDTreePCA`). This means the error bound is large so I tried retaining a higher amount of variance when performing PCA (e.g. 0.999999999) but the error bound is still too big, there's barely any difference with retaining 0.9 of the variance.

Are you sure we need to take into account query point reconstruction error? Note that I'm using `kernel(q, r) = kernel(q, qRecon) * kernel(qProj, rProj)`. I tried not taking into account query reconstruction error as well but almost same results :-1: .

And I don't think it's worth the overhead in the base case but I think it's the simplest place to code a first implementation and avoid issues that could come from working with trees, so that's why :)

Reply:
Oh, sorry, you're right---query reconstruction error is indeed accounted for with `kernelValue1`.  Sorry that I overlooked that.

So, if I understood correctly, you are saying that `metricProjValue - maxMetricError` is less than 0 for every single point (and that does not consider query reconstruction error).  And I think this can be true, but I also think we can be clever enough to work around this.

Take a look at Figure 1 (right side) in the paper.  We have `q`, `r`, `q_proj`, and `r_proj`, and our goal is to approximate `d(q, r)` quickly by projecting both `q` and `r` down to a lower-dimensional subspace where `q_proj` and `r_proj` are these projections.  By precomputing these projections and by the fact that the lower-dimensional distance computation is faster, our hope is that the extra error induced by these projections is small enough that we see speedup while still getting a kernel density estimate that matches our bounds.  (I think we're both on the same page about this, but for some reason I wrote it anyway.  Maybe because it is getting late.)

Specifically, for the sake of computing `K(q, r)`, we need to know `d(q, r)`---or an approximation thereof.  Since we are using the Gaussian kernel with the Euclidean distance, and since the vector `q -> q_proj` (also known as `q_recon`) and `q_proj -> r_proj` are orthogonal, it is true that `d(q, r_proj)^2 = d(q, q_proj)^2 + d(q_proj, r_proj)^2`.  This equality (not bound) here is how Dongryeol is able to decompose the distance computation in Eq. (3).  In fact, the "approximately equals" can be replaced with "exactly equals" if the left hand side is `e^{- || q - r_proj ||^2 / (2h^2)}`.  (I didn't double check that but I'm pretty sure.  Feel free to check...)

That means that the only error in the distance computation will come from the reference point reconstruction error.  Here, we don't have the nice orthogonality that we could use before, and so the best we can do is use the triangle inequality to say

```
d(q, r) <= d(q, r_proj) + d(r, r_proj)
```

where note that `d(r, r_proj)` is also equal to the (non-squared) norm of `r_recon`.  Similarly, we can use the triangle inequality in the other direction to say

```
d(q, r) >= | d(q, r_proj) - d(r, r_proj) |.
```

The important thing to notice here is that these bounds are on `d(q, r)`... but what we are using in the code is actually a bound on `d(q_proj, r)`.  Specifically, the code that uses `metricProjValue - maxMetricError` is equivalent to the bound `d(q_proj, r) > d(q_proj, r_proj) - d(r_proj, r)`; then, we later add in the orthogonal distance `d(q, q_proj)` via the computed kernel value `kernelValue1`.

So the first thing I notice is that we've actually forgotten part of the triangle inequality.  While it is true that `max(metricProjValue - maxMetricError, 0)` *is* a valid bound on `d(q_proj, r)`, we can actually do a bit better, because by the triangle inequality we should be able to just use `| metricProjValue - maxMetricError |` (or, `| d(q_proj, r_proj) - d(r, r_proj) |`).

The next observation is that the derivation we are using actually does not take into account the `d(q, q_proj)` term.  If we do that, we get a slightly different set of bounds on the distance:

```
d(q, r) <= sqrt(d(q, q_proj)^2 + d(q_proj, r_proj)^2) + d(r, r_proj)
d(q, r) >= | sqrt(d(q, q_proj)^2 + d(q_proj, r_proj)^2) - d(r, r_proj) |
```

For the upper bound, this doesn't really make *too* much difference, but I think that it can for the lower bound.  However, it's getting too late for me to think that through fully, so I'll give it another shot tomorrow.  Still though, I think that there are two things you can try:

(1) use `abs()` instead of `max(..., 0)` for the lower bound and see how that does
(2) use the more complex bounding strategy and don't compute `kernelValue1` directly

It's also possible to test whether the bounds are valid just by seeing if the actual kernel value is between `minK` and `maxK`---that should expose if something is horribly wrong in our thinking. :)

Reply:
I think your whole reasoning is correct :+1: 

Same issues when using (1), I will try (2) later tonight :)
Differences between the real value and the PCA calculated value are actually really big, e.g.
```
real vs pca
1.59032e-20 vs 9.4963e-14
1.1172e-19 vs 1.71036e-15
1.04255e-19 vs 9.75098e-14
4.10788e-16 vs 2.60875e-13.
2.66177e-19 vs 2.68426e-14
```
Also checked and real values are between `minK` and `maxK`, e.g.
```
1.59032e-20 -> 1.07165e-30 --- 1.96156e-10
1.1172e-19 -> 7.95175e-34 --- 5.08942e-10
1.04255e-19 -> 2.0276e-32 --- 9.9817e-11
4.10788e-16 -> 1.41317e-30 --- 8.89619e-11
2.66177e-19 -> 3.00812e-32 --- 2.65021e-10
```
So I guess the problem here is the range between `minK` and `maxK` is too large.

It might be possible that I misunderstood how PCA works, so an important thing to note here is that this only happens when some eigenvector is removed from the eigenvectors matrix (line 921 in `kde_impl.hpp`). If we retain 100% of variance then everything works fine.
Let's say my dataset points have 50 dimensions and PCA is computed in a leaf node of a KD-Tree, this means the eigenvector matrix will have dimension 50(rows)x(number of points in the node)(columns), so for a node that contains 8 points, dimension of the eigenvector matrix will be 50x8 (which means projections of points will have 8 dimensions). When we choose to retain 95% of variance this could mean that the last 2 columns are removed so the eigenvector matrix will be 50x6 and points projections have 6 dimensions. Is this right?

Reply:
Oh, sorry, maybe I should have looked more closely at the PCA code.  I'm about to go to bed so I don't have a chance now (I can do it tomorrow) but what you described is transposed---if our dataset has 50 dimensions, the goal of using PCA is to keep all 8 points, but project them onto a lower subspace.  For instance, if the majority of the variance of the data is contained in only a few dimensions, then, perhaps, only 5 dimensions are required to capture 95% of the variance; thus, after projection, we end up with 8 points in 5 dimensions, instead of 50.

When we do PCA, one way is to compute the eigendecomposition of the covariance matrix (which will be 50x50).  It sounds based on what you wrote like you are doing eigendecomposition of the data points, which is a slightly different thing.  (Actually you can use SVD on the data points themselves, but depending on the size of the data matrix, it may be slower or quicker, and the math to get the principal components is just a bit different.)

I'll take a better look at the code in the morning and hopefully be able to provide some better input then.

Reply:
Some updates: After implementing (2) it all seemed to work fine :tada: but then I tried to change dimension of the datasets as a test. It works fine for ~50 dimensions datasets but whenever we go to higher dimensionality (e.g. ~300 dimensions) I get a few estimations out of bound, and although they're usually few, this means there's still something wrong with the algorithm, so I'm trying to debug this.

Reply:
When you run on 300 dimensions (or more), are you doing any dimensionality reduction with PCA?  Also, how far out of bounds are the estimations?  (If you can print output or show me how to run it that could be helpful. :+1:)